### ISSUE TITLE
Add support for SAM/CloudFormation DefinitionSubstitutions

### ISSUE THREAD

**Issue by speshak** [ChristopheBougere/asl-validator#54]
When using [SAM DefinitionSubstitutions](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/sam-resource-statemachine.html#sam-statemachine-definitionsubstitutions) to provide Task Resource values, the validation fails because the token string doesn't match the ARN regex.

I realize this is hard to handle because the content of the replacement strings isn't known until deploy time.  Maybe this could be handled by a parameter that effectively means "DefinitionSubstitution tokens (`/\$\{.+\}/`) are always considered valid."  It does leave a bit of a validation hole, but it at least means that users of SAM/CFN can validate the rest of the ASL structure.

**Comment by ChristopheBougere** [comment]
Hi @speshak, sorry for the late answer.

There was a similar problem with a serverless plugin (serverless-step-functions) a while ago, see https://github.com/serverless-operations/serverless-step-functions/pull/245
This has been resolved by adding support for intrinsic functions within asl-validator: #35
That means that you can reference a resource either by an ARN or a `Ref` / `Fn:GetAtt` for instance ([see example here](https://github.com/ChristopheBougere/asl-validator/pull/35/files#diff-50f2bbfbce1eb4e39164f12fdd953f68R8))
Do you think this could solve your case?

If it doesn't, as this is an official SAM feature, I am fine adding it to `asl-validator`, would you be able to submit a PR?

**Comment by speshak** [comment]
I don't think that covers my usage.  I'm not using Ref/GetAtt because the ASL JSON isn't part of my CFN template, it's an external file that is pulled in by SAM using `DefinitionUri`.  I'll see about putting together a PR to cover this.

**Comment by ChristopheBougere** [comment]
OK, I understand better now, thanks for the explanation.
Yep, it might be hard to implement...

One easy way would be to add a new option with the regular expression `/\$\{.+\}/` [here](https://github.com/ChristopheBougere/asl-validator/blob/master/src/schemas/task.json#L25).
But this would only work for Resource attributes, and wouldn't handle all the cases:
>  Variables can be template parameter names, resource logical IDs, resource attributes, or a variable in a key-value map.

**Comment by ramgrandhi** [comment]
@ChristopheBougere - I'm looking to use `asl-validator` to validate my ASL JSON file. However, when used alongside `DefinitionSubstitutions` option, it results in validation failure.

Any developments on this thread?

```
Resource should match pattern "^arn:aws:([a-z]|-)+:([a-z]|[0-9]|-)*:[0-9]*:([a-z]|-)+:[a-zA-Z0-9-_.]+(:(\$LATEST|[a-zA-Z0-9-_]+))?$"
```

Cheers!

**Comment by ChristopheBougere** [comment]
Hi @ramgrandhi 
Nothing in progress that I am aware of, but PRs are welcome.

**Comment by ramgrandhi** [comment]
@ChristopheBougere  - can you please pick this PR up? Let me know if you're not okay with it. 

https://github.com/ChristopheBougere/asl-validator/pull/70

**Comment by ChristopheBougere** [comment]
Thanks for the PR @ramgrandhi 👍
Released in 1.9.6
